lib/PIClient/PrivacyIDEA.php

```diff
@@ -246,6 +247,130 @@ public function validateCheckWebAuthn(string $username, string $transactionID, s
+
+			// The userhandle and assertionclientextension fields are optional
+			if (!empty($passkeyResponseParams[USERHANDLE])) {
+				$params[USERHANDLE] = $passkeyResponseParams[USERHANDLE];
+			}
+			if (!empty($passkeyResponseParams[ASSERTIONCLIENTEXTENSIONS])) {
+				$params[ASSERTIONCLIENTEXTENSIONS] = $passkeyResponseParams[ASSERTIONCLIENTEXTENSIONS];
+			}
+
+			if (!empty($this->realm)) {
+				$params['realm'] = $this->realm;
+			}
+
+			$headers = $this->mergeHeaders($origin, $headers);
+			$response = $this->sendRequest($params, $headers, 'POST', '/validate/check');
+
+			return PIResponse::fromJSON($response, $this);
+		} else {
+			// Handle debug message if parameters are incomplete
+			$this->log('debug', 'validateCheckPasskey: parameters are invalid or incomplete!');
+		}
+		return null;
+	}
+
+	/**
+	 * Complete a passkey registration via the endpoint /validate/check. This is the second step of the registration process that was
+	 * triggered by the enroll_via_multichallenge setting in privacyIDEA.
+	 *
+	 * @param string $transactionID TransactionID.
+	 * @param string $serial Serial of the token.
+	 * @param string $username Username.
+	 * @param string $registrationResponse The registration data from the authenticator in json format.
+	 * @param string $origin Origin of the registrationResponse, usually gotten from a browser.
+	 * @param array|null $headers Optional headers for the request.
+	 * @return PIResponse|null PIResponse or null if error
+	 * @throws PIBadRequestException If an error occurs during the request.
+	 */
+	public function validateCheckCompletePasskeyRegistration(string $transactionID, string $serial, string $username, string $registrationResponse, string $origin, ?array $headers = null): ?PIResponse
+	{
+		assert(gettype($transactionID) === 'string');
+		assert(gettype($serial) === 'string');
+		assert(gettype($username) === 'string');
+		assert(gettype($registrationResponse) === 'string');
+		assert(gettype($origin) === 'string');
+
+		if (!empty($transactionID) && !empty($serial) && !empty($username) && !empty($registrationResponse) && !empty($origin)) {
+			try {
+				$registrationResponseParams = json_decode($registrationResponse, true);
+			} catch (\Exception $e) {
+				$this->log('debug', 'Invalid registration response for validateCheckCompletePasskeyRegistration: ' . $e->getMessage());
+				return null;
+			}
+			$params = ['transaction_id' => $transactionID, 'serial' => $serial, 'user' => $username, 'type' => 'passkey', CREDENTIAL_ID => $registrationResponseParams[CREDENTIAL_ID], CLIENTDATAJSON => $registrationResponseParams[CLIENTDATAJSON], ATTESTATIONOBJECT => $registrationResponseParams[ATTESTATIONOBJECT], AUTHENTICATORATTACHMENT => $registrationResponseParams[AUTHENTICATORATTACHMENT], RAWID => $registrationResponseParams[RAWID]];
```

Comment:
this line is way too long

Reply:
[02c2fcb](https://github.com/privacyidea/privacyidea-nextcloud-app/pull/26/commits/02c2fcbaa115d09edea04b08e21635b93a4ebe63)